bootloader/lk.go

```diff
@@ -60,38 +81,189 @@ func (l *lk) Name() string {
-	// at runtime environment is written directly into dedicated partition
-	if l.inRuntimeMode {
+	if l.prepareImageTime {
+		// at prepare-image time, then use rootdir and look for /boot/lk/ -
+		// this is only used in prepare-image time where the binary files exist
+		// extracted from the gadget
+		return filepath.Join(l.rootdir, "/boot/lk/")
+	}
+
+	// for runtime, we should only be using dir() for V1 and the dir is just
+	// the udev by-partlabel directory
+	switch l.role {
+	case RoleSole:
+		// TODO: this should be adjusted to use the kernel cmdline parameter
+		//       for the disk that the bootloader says to find the partition
+		//       on, since like the UC20 case, but that involves changing
+		//       many more tests, so let's do that in a followup PR
 		return filepath.Join(l.rootdir, "/dev/disk/by-partlabel/")
+	case RoleRecovery, RoleRunMode:
+		// TODO: maybe panic'ing here is a bit harsh...
+		panic("internal error: shouldn't be using lk.dir() for uc20 runtime modes!")
+	default:
+		panic("unexpected bootloader role for lk dir")
 	}
-	return filepath.Join(l.rootdir, "/boot/lk/")
 }
 
 func (l *lk) InstallBootConfig(gadgetDir string, opts *Options) error {
+	// make sure that the opts are put into the object
+	l.processOpts(opts)
 	gadgetFile := filepath.Join(gadgetDir, l.Name()+".conf")
-	systemFile := l.envFile()
+	systemFile, err := l.envFile()
+	if err != nil {
+		return err
+	}
 	return genericInstallBootConfig(gadgetFile, systemFile)
 }
 
 func (l *lk) Present() (bool, error) {
-	return osutil.FileExists(l.envFile()), nil
+	partitionLabelOrConfFile := l.partLabelForRoleAndTime()
+	// if we are not in runtime mode or in V1, just check the config file
+	if l.prepareImageTime || l.role == RoleSole {
+		return osutil.FileExists(filepath.Join(l.dir(), partitionLabelOrConfFile)), nil
+	}
+
+	// otherwise for V2, non-sole bootloader roles we need to check on the
+	// partition name existing, note that envFileForPartName will only return
+	// partiallyFound as true if it reasonably concludes that this is a lk
+	// device, so in that case forward err, otherwise return err as nil
+	_, partiallyFound, err := l.envFileForPartName(partitionLabelOrConfFile)
+	if partiallyFound {
+		return partiallyFound, err
+	}
+	return false, nil
 }
 
-func (l *lk) envFile() string {
+func (l *lk) partLabelForRoleAndTime() string {
+	// TODO: should the partition labels be fetched from gadget.yaml instead? we
+	//       have roles that we could use in the gadget.yaml structures to find
+	//       them
+	label := ""
+	switch l.role {
+	case RoleSole, RoleRunMode:
+		label = "snapbootsel"
+	case RoleRecovery:
+		label = "snaprecoverysel"
+	default:
+		panic(fmt.Sprintf("unknown bootloader role for littlekernel: %s", l.role))
+	}
+	if l.prepareImageTime {
+		// conf files at build time have .bin suffix, so technically this now
+		// becomes not a partition label but a filename, but meh names are hard
```

Comment:
all the callers have the if again, so maybe they should stick the suffix by themselves

Reply:
I think this is organized better now